conversion/conversion.go

```diff
@@ -165,12 +166,61 @@ func getDbNameFromSQLConnectionStr(driver, sqlConnectionStr string) string {
 	return ""
 }
 
+func getInfoSchemaForShard(shardConnInfo profiles.DirectConnectionConfig, driver string, targetProfile profiles.TargetProfile) (common.InfoSchema, error) {
+	params := make(map[string]string)
+	params["host"] = shardConnInfo.Host
+	params["user"] = shardConnInfo.User
+	params["dbName"] = shardConnInfo.DbName
+	params["port"] = shardConnInfo.Port
+	params["password"] = shardConnInfo.Password
+	sourceProfileConnectionMySQL, err := profiles.NewSourceProfileConnectionMySQL(params)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse connection configuration for the primary shard")
+	}
+	sourceProfileConnection := profiles.SourceProfileConnection{Mysql: sourceProfileConnectionMySQL, Ty: profiles.SourceProfileConnectionTypeMySQL}
+	//create a source profile which contains the sourceProfileConnection object for the primary shard
+	//this is done because GetSQLConnectionStr() should not be aware of sharding
+	newSourceProfile := profiles.SourceProfile{Conn: sourceProfileConnection, Ty: profiles.SourceProfileTypeConnection}
+	newSourceProfile.Driver = driver
+	infoSchema, err := GetInfoSchema(newSourceProfile, targetProfile)
+	if err != nil {
+		return nil, err
+	}
+	return infoSchema, nil
+}
+
 func schemaFromDatabase(sourceProfile profiles.SourceProfile, targetProfile profiles.TargetProfile) (*internal.Conv, error) {
 	conv := internal.MakeConv()
 	conv.SpDialect = targetProfile.Conn.Sp.Dialect
-	infoSchema, err := GetInfoSchema(sourceProfile, targetProfile)
-	if err != nil {
-		return conv, err
+	//handle fetching schema differently for sharded migrations, we only connect to the primary shard to
+	//fetch the schema. We reuse the SourceProfileConnection object for this purpose.
+	var infoSchema common.InfoSchema
+	var err error
+	switch sourceProfile.Ty {
+	case profiles.SourceProfileTypeConfig:
+		//Find Primary Shard Name
+		if sourceProfile.Config.ConfigType == "bulk" {
+			schemaShard := sourceProfile.Config.ShardConfigurationBulk.SchemaShard
```

Comment:
Nit: Probably change name from Schema shard to primaryShard, or something else without shard in the name ?

Reply:
Naming is hard.. 🥲
Renamed the variable and the object both to `schemaSource`.